text/0000-sensitive-data-redaction.md

```diff
@@ -0,0 +1,269 @@
+
+In this OTEP you will find a proposal for adding treatment of sensitive data to OpenTelemetry.
+
+By adding the proposed features, OpenTelemetry will provide its end-users the tooling needed to make sure that sensitive data is treated according to their requirements.
+
+This will make sure that these end-users can use OpenTelemetry within their secure and legal requirements and that OpenTelemetry (and implementing libraries) are able to avoid vulnerabilities.
+
+## Explanation
+
+### Overview
+
+This proposal aims to provide the following features:
+
+- methods for consumers of the OpenTelemetry API to enrich collected Attributes with sensitivity information and hooks to apply different levels of redaction.
+- related to those methods a similar way to enrich attributes in the semantic conventions with sensitivity information and ways of redaction.
+- a consistent way to configure sensitivity requirements for end-users of the OpenTelemetry SDK and in instrumentation libraries, including predefined “configuration profiles” and ways for fine grain configuration.
+- A redactor implements the logic to apply redaction and that owns predefined helpers for redaction in the SDK (URLParams filtering, Zeroing IPs, etc.).
+
+The following limitations apply:
+
+- Only Attributes are treated, although it is possible that sensitive data is contained in other data generated by OpenTelemetry as well (e.g. the span name, or the instrumentation scope name could contain sensitive data)
+
+### Annotate attributes with sensitivity information
+
+As a first building block the OpenTelemetry API needs to provide capability to enrich collected attributes with sensitivity details and with potential hooks to redact those. Those API changes then also can be used to describe sensitivity information in the semantic conventions.
+
+#### API
+
+Every API that sets an attribute consisting of a key and a value, needs to be enhanced by an additional functionality that allows to add details about the potential sensitivity of this data and a hooks how it may be redacted. This can be an additional set of parameters for an existing method or a method that can be called after the attribute has been set, if adding parameters to a method signature would lead to a breaking change.
+
+An additional method for setting the sensitivity information for a span attribute might look like the following:
+
+```
+span.setAttribute("url.query", url.toString(), <SENSITIVITY_DETAILS>);
```

Comment:
It's probably more of a spec concern and we'll have more time to design and polish API, but I'd like to entertain the idea of

- Some `SensitivityConfig` instrumentation can request from `Tracer`
- `SensitivityConfig` can be configured by the SDK/distro/etc and can have different implementations for different regulations
- `SensitivityConfig` has convenience methods to redact common things like URLs. E.g. `String SensitivityConfig.redactQueryParams(Uri uri)`
- It may recognize common attributes which are known to be problematic (e.g. `client.address` on the server spans is potentially PII) and could be applied implicitly without instrumentation code

E.g. 
```java
  Span clientSpan= tracer.startSpan("GET");
  String redactedUri = tracer.getSensitivityConfig().redactQueryParams(uri);
  clientSpan.setAttribute("url.full", redactedUri);
}
```

or 

```java
  Span serverSpan= tracer.startSpan("GET /foo", SERVER);
  // the `SensitivityConfig` associated with the tracer knows that `client.address` 
  // may be sensitive and would apply whatever it's configured to do 
  // allow, drop, anonymize, etc)
  serverSpan.setAttribute("client.address", request.getclientIP()); 
}
```

Reply:
I think it's both :)

I'd like to be able to add redaction/sanitization without **requiring** instrumentation to change their code or worry about it.

I.e. the alternative proposal is to build a reasonable story within existing API surface first.
But the proposal you made in the OTEL will still be necessary - we'll need to redact/sanitize custom attributes or to optimize and fine-tune instrumentation code for those who want it.

Reply:
Throwing in one more suggestion - we can auto-generate semantic convention code along with the redaction.

E,g, today we generate constants like:

```java
class UrlAttributes {
   public static final String URL_FULL = "url.full";
}
```

we could instead generate a method

```java
class UrlAttributes {
   public static final String URL_FULL = "url.full";
   public static void recordUrlFull(Span span, Uri uri) {
       String redacted = getSensitivityConfig().redactQueryParams(uri);
       span.setAttribute("url.full", redacted);
   }
}
```

We still need to have a central config and a way to pass it around and make it accessible on the API level, but we can hide a lot of boilerplate and be able to partially automate it.

/cc @lquerel

Reply:
> I'd like to be able to add redaction/sanitization without requiring instrumentation to change their code or worry about it.

Yes, that's an important feature! With me suggestions for how to write that down in the semantic conventions this should be feasible. Let me add some wording to make this clear.

Looking at the solutions I would prefer the first one over the updated code generation for the semantic conventions, i.e. `setAttribute` should do the redaction internally, because having a method per attribute seems excessive to me (and people **will** forget to use it, vs not using the constants comes with less penalties) and people can configure their own redaction through configuration (let's say the use an application they do not have code access and this application emits "enduser.id" without redaction, they can update their sensitivity config and make sure that it is redacted)

Reply:
One other problem with the suggestion. to specify redaction rules when the attribute is set is that every time someone defines an attribute, they have to consider redaction, which can fairly easily lead to forgetting about it and data leakage.

IMHO, I'd see a redaction system as something which runs in the background (@lmolkova's second suggestion), and can make redaction decisions based on various factors (the attribute name, or the value matching a regexp for example).

Reply:
I would suggest to start with the most generic solution that can address all leaks, even if it's less efficient. Anything that relies on instrumentation callsite changes is fundamentally not capable of full coverage - someone will always forget some sensitivity parameter or a call to sanitizer. Whereas having a config-driven sanitizer will be able to see all data and apply rules reliably.

Reply:
If you want to accomplish to address all leaks you need both: a config-driven sanitizer and one that can be used on instrumentation call site. Let me make an example:

Assume that `url.query` is sanitized as outlined in [971](https://github.com/open-telemetry/semantic-conventions/pull/971), so there is a list of terms that will be droppes, such that a query like `?q=OpenTelemetry&sig=Sp8h9C9xJotR` will turn into `?q=OpenTelemetry&sig=REDACTED`. Now imagine a library that communicates with a 3rd party service that the use a key for the signature that's not in that list (let's say X-Example-Signature), the implementor of that library needs a way to address that. Without an API interface for that the only option is going through the semantic conventions and waiting for it to be included + the languages adding it to their implementation, whereas an API interface allows them to do the following:

```
span.setAttribute(LANGUAGE_CONSTANT_FOR_URL_QUERY, theirValue, { DEFAULT => REDACT_INSECURE_URL_PARAM_VALUES AND values for X-Example-Signature })
```

Of course this requires some semantics to append (or overwrite) existing sensitivity details.

Also, thinking about what you say, if the "configuration-driven sanitizer" is also available to an instrumentor (means via the API), they could express their redaction needs in a centralized way as well:

```
// somewhere at library entry/setup/...
OpenTelemetry.registerSensitivityDetails({
  LANGUAGE_CONSTANT_FOR_URL_QUERY => { DEFAULT => REDACT_INSECURE_URL_PARAM_VALUES AND values for X-Example-Signature },
  /* other configurations that are relevant for the use of their library */
})
```

Reply:
I think this is already implicit, but to me, it's important to let users disable the redaction (ie, having a sensitivity details config that is permissive). This would allow them to not pay a performance penalty at runtime, and apply the redaction logic at the collector if this model satisfies their requirements.

Reply:
I reworked that section to lead with the configuration part (used some wording used here and in @pyohannes' suggestion), please take another look. I kept a section on the API but tried to express it's optional.

Reply:
> Without an API interface for that the only option is going through the semantic conventions and waiting for it to be included + the languages adding it to their implementation

The user deploying the library would already have the capability to redact any attribute produced in the telemetry through a central config. Why does it need to wait for anything in semantic conventions?

There is an aspect that the user may not know that there is an attribute that should be redacted, but that problem is solved by the schemas proposal (otel-weaver), where there is a standard way of introspecting all the shapes that instrumentation might produce.

Reply:
The user deploying that application has this capability through the configuration, yes, but what I was talking about is the library author only having the OpenTelemetry API available? How would they be able to express their redaction requirements? I see 3 options:

- They can also add configuration through the API, which is noop until the SDK is loaded. This has the advantage of them using the same logic as the user deploying the library.
- They can use a method to express their redactions. This gives them the most local control, i.e. when there is a specific place in their code where they want to have very specific redactions.
- They can not use OpenTelemetry for that and have their own logic for it. This is OK, if we want to keep that out of the conversation for now.

(related: https://github.com/open-telemetry/oteps/pull/255#discussion_r1593692898)

Reply:
Library authors do not have enough context to make reduction decisions. The best they can do is tag attributes with some semantic annotation saying "this is sensitive" (e.g. in my company we have a taxonomy of "purpose policy" annotations for data elements). If I deploy this library as a user, I should be able to decide if I actually want "sensitive" to be redacted or not - I may be debugging locally and need to see all data, or maybe in production I have different pathways where the data flows, some require reduction and others require raw data.

And to add these annotations it's better to go with the schemas approach, without changing the API.

Reply:
> And to add these annotations it's better to go with the schemas approach, without changing the API.

+1

Reply:
I think we are talking about the same thing, @yurishkuro: Indeed a library author can not make reduction decision, this is the responsibility of the application operator, but the library author needs capabilities to express that some data may be sensitive and provide redaction options similar to what we want to do in the semantic conventions. Here is an example: A library author writes a method that calls a payment provider HTTP endpoint, which has a unique query parameter to carry the token (let's say `acmeToken`). That library author now wants to say "when I call this endpoint, and a HTTP span is generated, make sure that `url.query` is redacted in a way that `acmeToken` is treated like other potentially sensitive attributes (see [#971](https://github.com/open-telemetry/semantic-conventions/pull/971)). Ideally they can do something similar to what is proposed for the semantic conventions, e.g. a way to write down what to do when `DEFAULT`, `STRICT` or `STRICTER` (or any other profile is selected). Then, the end user of that library and of opentelemetry can make the decision of redaction.

Reply:
@svrnm yes, this is a pretty complex use case, but I would still err on the side of a centralized solution over federating these responsibilities to the call sites. It's conceivable that the semantic convention / schema for `url.query` could be extended with conditional clauses like "if I am in this instrumentation scope or the URL contains this domain name, these are the additional rules / annotations for the query elements". And I think this approach is going to be necessary in many cases anyway, because the specific HTTP instrumentation may not have any clue about this business use case, e.g. it could be a generic HTTP client interceptor that generates client spans and records URL, and only user code would have the theoretical capability of understanding the sensitivity of the query params.

Reply:
> It's conceivable that the semantic convention / schema for url.query could be extended with conditional clauses like "if I am in this instrumentation scope or the URL contains this domain name, these are the additional rules / annotations for the query elements".

That's what is part of this proposal, I need to optimize the wording for that, but eventually the configuration for redaction needs to be that fine-grained.

> but I would still err on the side of a centralized solution over federating these responsibilities to the call sites.

Yes, centralized is the core of it, but library/application authors still need a way to provide a hint that a redaction needs to be applied. Rethinking my example, we actually have the following situation:

A library author is likely using a HTTP client library as dependency themselves to interact with a 3rd partyAPI, so in some pseudo code they might have something like the following:

```
function authenticateWithMyService($token) {
  // they want to let OpenTelemetry know that $token is sensitive, and when
  // it is injected into the `url.query` attribute certain redaction may be required.
  use3rdPartyHttpClient("https://myservice//?myAcmeToken=$token")
}
```

How would they implement that? They need some ways to annotate that. Also those annotations need to be composable somehow.